File: app/src/main/java/com/soundcloud/android/playback/flipper/FlipperWrapper.java
package com.soundcloud.android.playback.flipper;

import com.soundcloud.flippernative.api.Player;
import com.soundcloud.flippernative.api.PlayerListener;
import com.soundcloud.flippernative.api.audio_performance;
import com.soundcloud.flippernative.api.error_message;
import com.soundcloud.flippernative.api.state_change;

public class FlipperWrapper extends PlayerListener {

    private final Player flipper;
    private final FlipperAdapter flipperAdapter;

    public FlipperWrapper(FlipperAdapter flipperAdapter, FlipperFactory flipperFactory) {
        this.flipperAdapter = flipperAdapter;
        this.flipper = flipperFactory.create(this);
    }

    public void prefetch(String mediaUri) {
        flipper.prefetch(mediaUri);
    }

    public void play() {
        flipper.play();
    }

    public void pause() {
        flipper.pause();
    }

    public void seek(long positionMs) {
        flipper.seek(positionMs);
    }

    public float getVolume() {
        return (float) flipper.getVolume();
    }

    public void setVolume(float level) {
        flipper.setVolume(level);
    }

    public void destroy() {
        flipper.destroy();
    }

    public void open(String mediaUri, long positionMs) {
        flipper.open(mediaUri, positionMs);
    }

    void openEncrypted(String mediaUri, byte[] key, byte[] initVector, long positionMs) {
        flipper.openEncrypted(mediaUri, key, initVector, positionMs);
    }

    @Override
    public void onProgressChanged(state_change event) {
        flipperAdapter.onProgressChanged(new ProgressChange(event.getUri(), event.getPosition(), event.getDuration()));
    }

    @Override
    public void onPerformanceEvent(audio_performance event) {
        flipperAdapter.onPerformanceEvent(new AudioPerformanceEvent(event.getType().const_get_value(), event.getLatency().const_get_value(), event.getProtocol().const_get_value(),
                                                                    event.getHost().const_get_value(), event.getFormat().const_get_value(), (int) event.getBitrate().const_get_value(),
                                                                    event.getDetails().get_value().toJson()));
    }

    @Override
    public void onStateChanged(state_change event) {
        flipperAdapter.onStateChanged(mapToStateChange(event));
    }

    @Override
    public void onBufferingChanged(state_change event) {
        flipperAdapter.onBufferingChanged(mapToStateChange(event));
    }

    private StateChange mapToStateChange(state_change event) {
        return new StateChange(event.getUri(), event.getState(),
                               event.getReason(), event.getBuffering(),
                               event.getPosition(), event.getDuration(), event.getStreamingProtocol());
    }

    @Override
    public void onDurationChanged(state_change event) {
        // FIXME DO NOT CALL SUPER AS IT WILL CRASH THE APP WHILE SEEKING
        // FIXME Check JIRA: PLAYBACK-2706
    }

    @Override
    public void onSeekingStatusChanged(state_change stateChangeEvent) {
        flipperAdapter.onSeekingStatusChanged(new SeekingStatusChange(stateChangeEvent.getUri(), stateChangeEvent.getSeekingInProgress()));
    }

    @Override
    public void onError(error_message error) {
        flipperAdapter.onError(new FlipperError(error.getCategory(), error.getSourceFile(), error.getLine(),
                                                error.getErrorMessage(), error.getStreamingProtocol(),
                                                error.getCdn(), error.getFormat(), error.getBitRate()));
    }
}

File: app/src/main/java/com/soundcloud/android/playback/flipper/FlipperWrapper.kt
package com.soundcloud.android.playback.flipper

import com.soundcloud.android.playback.AudioPerformanceEvent
import com.soundcloud.android.playback.PlaybackMetric
import com.soundcloud.android.utils.OpenForTesting
import com.soundcloud.flippernative.api.PlayerListener
import com.soundcloud.flippernative.api.audio_performance
import com.soundcloud.flippernative.api.error_message
import com.soundcloud.flippernative.api.state_change

@OpenForTesting
internal class FlipperWrapper
constructor(private val flipperCallbacks: FlipperCallbacks, flipperFactory: FlipperFactory)
    : PlayerListener() {

    private val flipper = flipperFactory.create(this)

    fun prefetch(mediaUri: String) = flipper.prefetch(mediaUri)
    fun play() = flipper.play()
    fun pause() = flipper.pause()
    fun seek(positionMs: Long) = flipper.seek(positionMs)
    fun destroy() = flipper.destroy()
    fun open(mediaUri: String, positionMs: Long) = flipper.open(mediaUri, positionMs)
    fun openEncrypted(mediaUri: String, key: ByteArray, initVector: ByteArray, positionMs: Long) = flipper.openEncrypted(mediaUri, key, initVector, positionMs)

    var volume: Double
        get() = flipper.volume
        set(level) {
            flipper.volume = level
        }

    override fun onProgressChanged(event: state_change) {
        flipperCallbacks.onProgressChanged(ProgressChange(event.uri, event.position, event.duration))
    }

    override fun onPerformanceEvent(event: audio_performance) {
        flipperCallbacks.onPerformanceEvent(AudioPerformanceEvent(PlaybackMetric.from(event.type.const_get_value()), event.latency.const_get_value(), event.protocol.const_get_value(),
                                                                  event.host.const_get_value(), event.format.const_get_value(), event.bitrate.const_get_value().toInt(),
                                                                  event.details._value.toJson()))
    }

    override fun onStateChanged(event: state_change) {
        flipperCallbacks.onStateChanged(mapToStateChange(event))
    }

    override fun onBufferingChanged(event: state_change) {
        flipperCallbacks.onBufferingChanged(mapToStateChange(event))
    }

    private fun mapToStateChange(event: state_change): StateChange {
        return StateChange(event.uri, event.state,
                           event.reason, event.buffering,
                           event.position, event.duration, event.streamingProtocol)
    }

    override fun onDurationChanged(event: state_change) {
        // FIXME DO NOT CALL SUPER AS IT WILL CRASH THE APP WHILE SEEKING
        // FIXME Check JIRA: PLAYBACK-2706
    }

    override fun onSeekingStatusChanged(stateChangeEvent: state_change) {
        flipperCallbacks.onSeekingStatusChanged(SeekingStatusChange(stateChangeEvent.uri, stateChangeEvent.seekingInProgress))
    }

    override fun onError(error: error_message) {
        flipperCallbacks.onError(FlipperError(error.category, error.sourceFile, error.line,
                                              error.errorMessage, error.streamingProtocol,
                                              error.cdn, error.format, error.bitRate))
    }
}
